### PR TITLE
:sparkles: Add incarnation

### PR DESCRIPTION
**Pull request by tongzhi1998**
Problem:
- If an account is deleted and created again in the same block, we need to remove the old storage completely
- Our current code doesn't support for that

Solution:
- Add incarnation to Account 
- If it is a new account, then remove old storage from the db

### CODE REVIEW

File: include/monad/db/in_memory_trie_db.hpp [hunk 1]
```diff
@@ -74,6 +74,9 @@ struct InMemoryTrieDB : public Db
         address_t const &a, uint64_t incarnation,
         bytes32_t const &key) const override
     {
+        if (MONAD_UNLIKELY(incarnation != 0)) {
+            return bytes32_t{};
+        }
```
Comment (zander-xyz) [inline include/monad/db/in_memory_trie_db.hpp]: i would have thought that reads for `incarnation != 0` never makes it to the db layer

File: include/monad/db/rocks_trie_db.hpp [hunk 2]
```diff
@@ -215,6 +216,9 @@ struct RocksTrieDB : public Db
         address_t const &a, uint64_t incarnation,
         bytes32_t const &key) const override
     {
+        if (MONAD_UNLIKELY(incarnation != 0)) {
+            return bytes32_t{};
+        }
```
Comment (zander-xyz) [inline include/monad/db/rocks_trie_db.hpp]: I think we should instead assert that incarnation is zero. We should never call into the DB if incarnation is zero - that should be handled inside of read_*

File: include/monad/db/trie_db_process_changes.hpp [hunk 3]
```diff
@@ -129,6 +129,9 @@ void trie_db_process_changes(
         // process storage updates first
         storage_trie_updates.clear();
         if (account_delta.second.has_value()) {
+            if (MONAD_UNLIKELY(account_delta.second.value().incarnation) != 0) {
```
Comment (zander-xyz) [inline include/monad/db/trie_db_process_changes.hpp]: this is not necessarily unlikely

also, I think we should also check against `orig.has_value()`. practically it should not matter but it doesn't make sense to try to clear the storage trie if it never existed to start.

did you test what happens if, in the same set of transactions, the account gets destructed and recreated multiple times? does this logic break down?

File: include/monad/state2/state.hpp [hunk 4]
```diff
@@ -209,13 +216,19 @@ struct State
 
     void destruct_suicides() const noexcept {}
 
+    bool is_dead(Account const &account)
+    {
+        return account.balance == 0 && account.nonce == 0 &&
+               account.code_hash == NULL_HASH;
+    }
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp]: can you please rebase on top of main? my changes add this function

File: include/monad/state2/block_state.hpp [hunk 5]
```diff
@@ -12,6 +12,7 @@ struct BlockState
     Mutex mutex;
     StateDeltas state;
     Code code;
+    Incarnations incarnations;
```
Comment (zander-xyz) [inline include/monad/state2/block_state.hpp]: why is this needed?

File: include/monad/trie/rocks_writer.hpp [hunk 6]
```diff
@@ -54,6 +55,15 @@ struct RocksWriter
         auto const res = batch.DeleteRange(
             cf, monad::db::to_slice(buf.view()), monad::db::to_slice(end));
         MONAD_ROCKS_ASSERT(res);
+
+        write();
```
Comment (zander-xyz) [inline include/monad/trie/rocks_writer.hpp]: It's fine to immediately write the del prefix, but can we do this at the call-site instead of right here? That seems to match the current usage a bit more intuitively

File: include/monad/db/in_memory_trie_db.hpp (hunk 1 above)
Comment (tongzhi1998) [reply include/monad/db/in_memory_trie_db.hpp]: That makes sense. However, I think we still want to push `bytes32_t{}` to both blockstate and state, so the current logic in `read_storage` seems appropriate. Moving this to the read_storage() end might introduce some redundent code.

File: include/monad/db/trie_db_process_changes.hpp (hunk 3 above)
Comment (tongzhi1998) [reply include/monad/db/trie_db_process_changes.hpp]: Cool, I can do that. I will also add test case for multiple deletions & creations.

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: Because we need to record how many times an account is being destructed. Upon self-destruct, account is blockstate is being reset so it loses the incarnation value that it holds

File: include/monad/db/in_memory_trie_db.hpp (hunk 1 above)
Comment (zander-xyz) [reply include/monad/db/in_memory_trie_db.hpp]: why is it redundant?

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (zander-xyz) [reply include/monad/state2/block_state.hpp]: why do you need to keep track how many times?

File: include/monad/db/in_memory_trie_db.hpp (hunk 1 above)
Comment (tongzhi1998) [reply include/monad/db/in_memory_trie_db.hpp]: Sorry I was misunderstanding this before. Fixed now!

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: Might be a bit complicated to explain. Maybe we can zoom to discuss.

Comment (zander-xyz) [reply include/monad/state2/block_state.hpp]: Can you present a simple scenario where you need to keep track of the number of times the account has reincarnated?

File: include/monad/db/trie_db_process_changes.hpp [hunk 7]
```diff
@@ -129,6 +129,11 @@ void trie_db_process_changes(
...
```
Comment (zander-xyz) [inline include/monad/db/trie_db_process_changes.hpp]: sorry, I think my comment was confusing
> this is not necessarily unlikely

this was suggesting that we get rid of MONAD_UNLIKELY

> also, I think we should also check against orig.has_value(). practically it should not matter but it doesn't make sense to try to clear the storage trie if it never existed to start.

did this get addressed?

File: include/monad/state2/state.hpp [hunk 8]
```diff
@@ -77,6 +77,7 @@ struct State
         return account.has_value();
     };
 
+    // TODO: Lock Issue
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp]: what's this?

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: I can try. Let's say a block has 5 txns (in order):
1: destroy account A
2: re-create account A (with storage)
3: destroy account A again
4: re-create account A again (with diff storage)
5. Access account A's storage

In optimistic execution, let's say 5 run after 2, followed by 3 & 4. Without incarnation number, 5 will be able to merge, whereas in reality, 5 can't be merged.

Comment (zander-xyz) [reply include/monad/state2/block_state.hpp]: Incarnation number is on the account, so would this not fail to merge (that is done in order)? In your example above, for each transaction, these would be the incarnations.

| transaction | `A` incarnation (before, after) |
-----|--------
| 1     |  (0, 0)        |
| 2     |  (0, 1)        |
| 5     |  (1, 1)        |
| 3     |  (1, 1)        |
| 4     |  (1, 2)        |

but then we merge in order,

| transaction | `A` incarnation (before, after) |
-----|--------
| 1     |  (0, 0)        |
| 2     |  (0, 1)        |
| 3     |  (1, 1)        |
| 4     |  (1, 2)        |
| 5     |  (1, 1)        |

when trying to merge the state deltas for 5 into 4 (which includes 1, 2, 3, 4), we should see that the orig for 5 != updated for 4, and fail to merge.

Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: Incarnation number is not on the account (because of reset). Incarnation bool can be on the account, but the above illustrates why a boolean will not be enough.

Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: Only by adding this incarnation map can incarnation number be on the account.

Comment (zander-xyz) [reply include/monad/state2/block_state.hpp]: Sorry, can you link me to the reset you are referring to?

File: include/monad/state2/state.hpp (hunk 8 above)
Comment (tongzhi1998) [reply include/monad/state2/state.hpp]: This function need to be protected by a lock since we are accessing blockstate directly here. Also I think even w/o the current implementation of accessing blockstate, we need a lock to protect the whole function. 
From my understanding, when EVMOne will cause create_account when it thinks an account doesn't exist, but if we don't lock this function, upon executing read_account, the account might actually be in the blockstate already.

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: Account is resetted upon self-destruct, and merged into blockstate. Upon merging, in blockstate, account.current = std::nullopt, and therefore lose the incarnation that it previous has

File: include/monad/db/trie_db_process_changes.hpp (hunk 7 above)
Comment (tongzhi1998) [reply include/monad/db/trie_db_process_changes.hpp]: Sorry I think I forget to address this one. I will do it in a sec.

File: include/monad/state2/block_state.hpp (hunk 5 above)
Comment (tongzhi1998) [reply include/monad/state2/block_state.hpp]: The map is there so that every time on reset, it records the previous incarnation that an account has so that we don't lose track of it.

File: include/monad/db/trie_db_process_changes.hpp (hunk 7 above)
Comment (tongzhi1998) [reply include/monad/db/trie_db_process_changes.hpp]: I think 2 conditions combined will make it unlikely. So I will keep it for now.

File: include/monad/db/in_memory_trie_db.hpp [hunk 9]
```diff
@@ -51,6 +51,23 @@ struct InMemoryTrieDB : public Db
...
+            trie_cursor.reset();
+        }
+
+        void clear_and_reset_prefix(address_t const &addr)
+        {
+            trie.clear();
+            trie.leaves_writer_.write();
+            trie.trie_writer_.write();
+
+            reset_cursor();
```
Comment (zander-xyz) [inline include/monad/db/in_memory_trie_db.hpp]: delete this

File: src/monad/state2/state_deltas.cpp [hunk 10]
```diff
@@ -57,9 +57,14 @@ void merge(StateDeltas &s1, StateDeltas const &s2)
...
-        merge_f(d1.storage, d2.storage, [](auto &st1, auto const &st2) {
-            st1.second = st2.second;
-        });
+        if (d2.account.second.has_value()) {
+            merge_f(d1.storage, d2.storage, [](auto &st1, auto const &st2) {
+                st1.second = st2.second;
+            });
+        }
+        else {
+            d1.storage.clear();
```
Comment (zander-xyz) [inline src/monad/state2/state_deltas.cpp:66]: now that we do this, we should probably `MONAD_DEBUG_ASSERT(!updated_storage_roots.contains(addr))` in `trie_db_process_changes` instead of `updated_storage_roots.erase(addr);`

File: include/monad/db/in_memory_trie_db.hpp (hunk 9 above)
Comment (tongzhi1998) [reply include/monad/db/in_memory_trie_db.hpp]: Makes sense.
